src/python/pants/backend/nfpm/native_libs/deb/search_for_sonames.py

```diff
@@ -0,0 +1,209 @@
+# Copyright 2025 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+"""A standalone script that uses asyncio to find packages that contain a SONAME.
+
+Rule code should run this script as a subprocess in a sandboxed venv. Generally, that venv should
+be a subset of the pants venv (using the version of packages that pants depends on).
+
+WARNING: This script relies on aiohttp and asyncio, so their are probably incompatibilities
+with rule code in the pants engine. So, avoid importing this script in rule code.
+"""
+
+from __future__ import annotations
+
+import argparse
+import asyncio
+import json
+import sys
+from collections import defaultdict
+from collections.abc import Generator, Iterable
+
+import aiohttp
+import aiohttp_retry
+from aiohttp.http import SERVER_SOFTWARE as DEFAULT_USER_AGENT
+from aiohttp_retry.types import ClientType
+from bs4 import BeautifulSoup
+
+DISTRO_PACKAGE_SEARCH_URL = {
+    "debian": "https://packages.debian.org/search",
+    "ubuntu": "https://packages.ubuntu.com/search",
+}
+
+
+async def deb_search_for_sonames(
+    distro: str,
+    distro_codename: str,
+    debian_arch: str,
+    sonames: Iterable[str],
+    user_agent: str = DEFAULT_USER_AGENT,
+) -> dict[str, dict[str, list[str]]]:
+    """Given a soname, lookup the deb package that provides it.
+
+    Tools like 'apt-get -S' and 'apt-file' only work for the host's active distro and distro
+    version. This code, however, should be able to run on any host even non-debian and non-ubuntu
+    hosts. So, it uses an API call instead of local tooling.
+    """
+    search_url = DISTRO_PACKAGE_SEARCH_URL[distro]
+
+    # tasks are IO bound
+    async with (
+        aiohttp_retry.RetryClient(
+            retry_options=aiohttp_retry.JitterRetry(attempts=5),
+            headers={aiohttp.hdrs.USER_AGENT: user_agent},
+            # version=aiohttp.HttpVersion11,  # aiohttp does not support HTTP/2 (waiting for contribution)
+            # timeout=aiohttp.ClientTimeout(total=5 * 60, sock_connect=30),
+        ) as client,
```

Comment:
The script accepts a `--user-agent-suffix` so that we can make sure debian knows that this traffic is coming from pants. Hopefully if this causes any issues, they can reach out and ask for a different approach. Basically, my philosophy is to make sure I am blame-able (git blame is beautiful!) when I change things, to speed up the process of telling me about [inadvertent] issues so I can fix them ASAP.

Also, I have a `timeout` stubbed in here. The `timeout` arg (and `headers`) gets passed to an `aiohttp.ClientSession`. The commented value represents the default timeout (5 minutes with sock_connect timeout of 30 seconds). Should I just remove this comment, or do something else with it?